imports/methods/TypedMethod.ts

```diff
@@ -0,0 +1,67 @@
+import { check } from 'meteor/check';
+import { EJSONable, EJSONableProperty } from 'meteor/ejson';
+import { Meteor } from 'meteor/meteor';
+
+type TypedMethodParam = EJSONable | EJSONableProperty
+type TypedMethodArgs = Record<string, TypedMethodParam>;
+type TypedMethodValidator<Arg extends TypedMethodArgs> =
+  (this: Meteor.MethodThisType, arg0: Record<keyof Arg, unknown>) =>
+    Arg;
+type TypedMethodRun<Arg extends TypedMethodArgs, Return extends TypedMethodParam | void> =
+  (this: Meteor.MethodThisType, arg0: Arg) => Return;
+
+export default class TypedMethod<
+  Args extends TypedMethodArgs,
+  Return extends TypedMethodParam | void,
+> {
+  name: string;
+
+  validate: TypedMethodValidator<Args>;
+
+  run: TypedMethodRun<Args, Return>;
+
+  constructor({ name, validate, run }: {
+    name: string,
+    validate: TypedMethodValidator<Args>,
+    run: TypedMethodRun<Args, Return>,
+  }) {
+    check(name, String);
+    check(validate, Function);
+    check(run, Function);
+
+    this.name = name;
+    this.validate = validate;
+    this.run = run;
+
+    /* eslint-disable-next-line @typescript-eslint/no-this-alias -- we need to
+       capture both the TypedMethod object (so we can find the implementation)
+       and the method invocation context (for things like userId) */
+    const self = this;
+    Meteor.methods({
+      /* eslint-disable-next-line meteor/audit-argument-checks -- because we're
```

Comment:
Just to make sure I understand correctly -- you're saying the audit-argument-checks lint will no longer fire, but the runtime checks will still complain serverside, throwing a Meteor.Error if we try to access fields without first check()-ing them?

Reply:
Yeah, that's right. `audit-argument-checks` exists as both a linter and a Meteor package. The linter statically inspects any method that's passed to `Meteor.methods` to ensure that variables get `check`'d. That's not useful here since we're effectively establishing a wrapper. However, disabling the linter doesn't do anything to the (older) runtime check, which will still fire based on the presence of the Meteor package `audit-argument-checks`.

(It would be _optimal_ if we could (say) apply the linter to the `validate` method of `TypedMethod`, but the linter is written pretty narrowly, so I don't think there's any way to plug that in.)

Reply:
Hmm. It turns out I did not entirely understand this correctly. The `audit-argument-checks` Meteor package does object identity-based tracking of [if you've looked at the top-level argument](https://github.com/meteor/meteor/blob/9530de1ac97f7736bfa7caeffda6a756603919fd/packages/check/match.js#L470-L517), so checking a destructured object doesn't count. So we can't use both destructuring and `audit-argument-checks`, which is unfortunate.

Reply:
Ah, so to be clear -- we'd have to turn off `audit-argument-checks` for the whole program to use destructuring anywhere?  I guess that's still fine if we don't mind doing the migration all-at-once.  I guess the other option is not using destructuring in `validate()` and `check()`ing the top-level object mid-migration, but if we do that I'd want to be extra careful about rejecting unknown keys.